src/qibolab/_core/instruments/qrng.py

```diff
@@ -0,0 +1,151 @@
+from typing import Iterable, Optional, Union
+
+import numpy as np
+import numpy.typing as npt
+from scipy.linalg import toeplitz
+from serial import Serial
+
+from .abstract import Instrument
+
+__all__ = ["QRNG"]
+
+
+def unpackbits(x: npt.NDArray, num_bits: int) -> npt.NDArray:
+    if np.issubdtype(x.dtype, np.floating):
+        raise ValueError("numpy data type needs to be int-like")
+    xshape = list(x.shape)
+    x = x.reshape([-1, 1])
+    mask = 2 ** np.arange(num_bits, dtype=x.dtype).reshape([1, num_bits])
+    return (x & mask).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+
+def generate_toeplitz(input_bits: int, extraction_ratio: int) -> npt.NDArray:
+    """Generate a pseudo-random Toeplitz matrix of dimension ``(input_bits, extraction_ratio)``."""
+    while True:
+        c = np.mod(np.random.permutation(input_bits), 2)
+        r = np.mod(np.random.permutation(extraction_ratio), 2)
+        if np.sum(c) == 6 and np.sum(r) == (extraction_ratio // 2):
+            return toeplitz(c, r)
+
+
+def extractor(m1: npt.NDArray, input_bits: int, extraction_ratio: int) -> npt.NDArray:
+    """Extract uniformly distributed integers from the device samples."""
+    m2 = unpackbits(m1, input_bits)
+    m2 = np.flip(m2)
+
+    m3 = generate_toeplitz(input_bits, extraction_ratio)
+
+    m4 = np.matmul(m2, m3)
+    m4 = m4.astype("int16")
+    m4 = np.mod(m4, 2)
+    m4 = np.packbits(m4)
+    m4 = np.right_shift(m4, 8 - extraction_ratio)
+    return m4
+
+
+def upscale(samples: npt.NDArray, input_bits=4, output_bits=32) -> npt.NDArray:
+    """Increase size of random bit strings by concatenating them."""
+    assert output_bits > input_bits
+    assert output_bits % input_bits == 0
+
+    factor = output_bits // input_bits
+    assert len(samples) % factor == 0
+
+    n = len(samples) // factor
+    upscaled = np.zeros(n, dtype=int)
+    for i, s in enumerate(np.reshape(samples, (factor, n))):
+        upscaled += s << (input_bits * i)
+    return upscaled
+
+
+class QRNG(Instrument):
+    """Driver to sample numbers from a Quantum Random Number Generator (QRNG).
+
+    Note that if we are not connected to a physical QRNG device, this will
+    return pseudo-random generated numbers using ``numpy.random``.
```

Comment:
Is it actually worth it? Isn't it better to just return an error, as it happens for other devices when it is impossible to connect?

Reply:
One potential benefit for users is that they can easily switch between numpy pseudo-random and QRNG quantum-random by commenting out `qrng.connect()` and check if there is any difference in their application. But there are probably other ways this can be done.

Another motivation was that we can get a bit higher coverage from the CI once I write tests, but I guess this could also be achieved with mocking.

Reply:
> One potential benefit for users is that they can easily switch between numpy pseudo-random and QRNG quantum-random by commenting out `qrng.connect()` and check if there is any difference in their application. But there are probably other ways this can be done.

I agree about the relevance of the comparison, but I'd keep it at a higher level than the driver.

> Another motivation was that we can get a bit higher coverage from the CI once I write tests, but I guess this could also be achieved with mocking.

I acknowledge we have a coverage issue. But yes, mocking here is the way, otherwise better the poor coverage (since it's just true that we are lacking the conditions to test in the CI).

Reply:
In af710eb1dda711a5ba05502c2f51a4c066bde998 I dropped this "feature" and replaced with mocking in tests, so I will resolve this.